### ISSUE TITLE
vg deconstruct

### ISSUE THREAD

**Issue by mahulchak** [vgteam/vg#800]
Hello,

I am trying to run vg deconstruct on a vg file that I created from whole genome alignments. However, the command did not generate any output. None of the subcommands worked for deconstruct (a few other vg commands I tried are working). Any idea why it might not be working? 

Thanks.
Mahul

**Comment by edawson** [comment]
Hey Mahul,

I broke vg deconstruct a few months back when we moved from one bubble detection algorithm to another. I haven't taken the time to fix it, but it has been on my todo list.

I will try to get to it this week, as you are not the first to attempt using it. In the meantime, `call` is a viable variant caller, and `snarls` can produce some useful output for looking at bubbles in the graph.

Best,
    Eric

**Comment by mahulchak** [comment]
Hi Eric,
Thanks for replying quickly. I don't have a read alignment or .gam file so
I was not sure if I could use the call subcommand. Deconstruct might be the
most appropriate subcommand for me but I could be wrong. I'll play with
call and snarls and see what I get.
Hopefully you'll have time to reactivate deconstruct ☺
Thanks,
Mahul

**Comment by edawson** [comment]
Github automatically closed this after my PR merged. I've reopened it till we get some verification it at least partially works.

@mahulchak , if your vg file you've created (I assume with MSGA) has paths, then deconstruct will be able to pull out one of these paths at a time (specified on the command line) and give you a VCF containing variation relative to that path. Is that what you'd like?

I have merged in a new version of deconstruct that *should* work for SNPs and smallish events. Complex structural variation will quite possibly break it.

Mind giving it a try?

P.S.: If you don't have any paths things get a bit more complicated - orienting on a graph without paths is weird. We might still be able to do it but it won't make a proper VCF.

**Comment by mahulchak** [comment]
Hi Eric,
Thanks again. I'm working on a test dataset of 1Mb chunk from 15 genomes.
I'll test on it later today or early tomorrow and let you know.
Mahul

**Comment by mahulchak** [comment]
Naive question: Does the --path option in deconstruct refer to
genome/sequence names? I tried using names of one of the genomes as well as
one of the ancestral genome (Anc0) but I am getting -

vg: src/path_index.cpp:254: vg::PathIndex::PathIndex(vg::VG&, const
string&, bool): Assertion `vg.paths.has_path(path_name)' failed.

Am I missing an argument?

**Comment by edawson** [comment]
The `-p / --path` option should refer to a path in the graph, yes. You can get a list of all the path by doing `vg paths -L <my_graph>.vg`, which should print a list of all the paths labeled in the graph. If your path isn't listed then it isn't labeled in the graph for some reason.

**Comment by edawson** [comment]
@mahulchak I've confirmed with the team that `vg msga ` *should* incorporate and label your paths; they should show up in `vg paths -L ` and should work for deconstruct.

**Comment by mahulchak** [comment]
Yes the paths are named after the chromosomes which were changed during hal
to vg conversion. I now have the names of the paths and testing vg
deconstruct for one of the paths.

**Comment by edawson** [comment]
I'm thinking that's a bug on my end, but it's hard to debug without a test case. If you want to share data via email (mine's on my Github profile) that would help; otherwise I can try to generate one.

**Comment by mahulchak** [comment]
I will share the vg file with you shortly.

**Comment by mahulchak** [comment]
Did you get the vg file?

**Comment by mahulchak** [comment]
It is crashing due to excessive memory use. E.g. My vg file has alignment of 15 1Mb chunks and _vg deconstruct_ is taking more than 300GB memory.

**Comment by mahulchak** [comment]
I tried aligning the same 15 1Mb chunks of 15 genomes using _vg msga_ and then tried _vg deconstruct_ on the vg file. After getting 15 lines written into the vcf file, I get - 
terminate called after throwing an instance of 'std::runtime_error'
  what():  No node 0 in graph

**Comment by ekg** [comment]
Have we moved over to snarls for deconstruct yet?

**Comment by edawson** [comment]
Yes, but it's having memory usage issues apparently. It works on the test cases in vg.

**Comment by mahulchak** [comment]
If you want, I have another vg file based an even smaller (100Kb) and
simpler sequences. And I have a vg made with vg msga and another created
from a hal file. vg deconstruct crashes due to excessive memory usage for
both vg files.

**Comment by edawson** [comment]
@mahulchak could you send your smaller test case? My guess is that I'm caching something that I've underestimated the size of.

**Comment by ChriKub** [comment]
@edawson I get the std:bad_alloc as well. Is it a new bug or does the old one still persist?
I could provide you with a small test case if it helps.
Thanks, 
Chris
